autoarray/plot/mat_wrap/wrap/wrap_base.py

```diff
@@ -0,0 +1,946 @@
+        """
+        if isinstance(self.config_dict["aspect"], str):
+            if self.config_dict["aspect"] in "square":
+                return float(shape_2d[1]) / float(shape_2d[0])
+
+        return self.config_dict["aspect"]
+
+    def open(self):
+        """Wraps the Matplotlib method 'plt.figure' for opening a figure."""
+        if not plt.fignum_exists(num=1):
+            plt.figure(**self.config_dict_figure)
+
+    def close(self):
+        """Wraps the Matplotlib method 'plt.close' for closing a figure."""
+        if plt.fignum_exists(num=1):
+            plt.close()
+
+
+class Cmap(AbstractMatWrap):
+    def __init__(self, **kwargs):
+        """
+        Customizes the Matplotlib colormap and its normalization.
+
+        This object wraps the following Matplotlib methods:
+
+        - colors.Linear: https://matplotlib.org/3.3.2/tutorials/colors/colormaps.html
+        - colors.LogNorm: https://matplotlib.org/3.3.2/tutorials/colors/colormapnorms.html
+        - colors.SymLogNorm: https://matplotlib.org/3.3.2/api/_as_gen/matplotlib.colors.SymLogNorm.html
+
+        The cmap that is created is passed into various Matplotlib methods, most notably imshow:
+
+         https://matplotlib.org/3.3.2/api/_as_gen/matplotlib.pyplot.imshow.html
+        """
+
+        super().__init__(kwargs=kwargs)
+
+    def norm_from_array(self, array: np.ndarray) -> object:
+        """
+        Returns the `Normalization` object which scales of the colormap.
+
+        If vmin / vmax are not manually input by the user, the minimum / maximum values of the data being plotted
+        are used.
+
+        Parameters
+        -----------
+        array : np.ndarray
+            The array of data which is to be plotted.
+        """
+
+        if self.config_dict["vmin"] is None:
+            vmin = np.min(array)
+        else:
+            vmin = self.config_dict["vmin"]
+
+        if self.config_dict["vmax"] is None:
+            vmax = np.max(array)
+        else:
+            vmax = self.config_dict["vmax"]
+
+        if self.config_dict["norm"] in "linear":
+            return colors.Normalize(vmin=vmin, vmax=vmax)
+        elif self.config_dict["norm"] in "log":
+            if vmin == 0.0:
+                vmin = 1.0e-4
+            return colors.LogNorm(vmin=vmin, vmax=vmax)
+        elif self.config_dict["norm"] in "symmetric_log":
+            return colors.SymLogNorm(
+                vmin=vmin,
+                vmax=vmax,
+                linthresh=self.config_dict["linthresh"],
+                linscale=self.config_dict["linscale"],
+            )
+        else:
+            raise exc.PlottingException(
+                "The normalization (norm) supplied to the plotter is not a valid string (must be "
+                "{linear, log, symmetric_log}"
+            )
+
+
+class Colorbar(AbstractMatWrap):
+    def __init__(
+        self,
+        manual_tick_labels: typing.Union[typing.List[float]] = None,

```

Comment:
Optional not Union

Union is for when the type could be one of a set of two or more types

Reply:
done